### ISSUE TITLE
Residual Unit implementation in tensorflow

### ISSUE THREAD

**Issue by issa-s-ayoub** [tensorflow/models#4505]
I am trying to implement the Residual Unit of Resnet v1 and v2 based on https://arxiv.org/pdf/1512.03385.pdf and https://arxiv.org/pdf/1603.05027.pdf. So here are the residual units that I am concerned about implementing in tensorflow:

![image](https://user-images.githubusercontent.com/26584101/41238539-2497158c-6d64-11e8-9608-a35f18c23dac.png)

And the v2 Residual Unit is (b) below:

![image](https://user-images.githubusercontent.com/26584101/41238554-2ed6653e-6d64-11e8-8fdd-ff95b71167e2.png)

Hence, I was looking into their implementation in https://github.com/tensorflow/models/blob/master/official/resnet/resnet_model.py and found the following for resnet unit v1:

	def _building_block_v1(inputs, filters, training, projection_shortcut, strides,
						   data_format)

		if projection_shortcut is not None:
			shortcut = projection_shortcut(inputs)
			shortcut = batch_norm(inputs=shortcut, training=training, data_format=data_format)

so my first question is: shouldn't we use the batch norm after the addition and never mess up with the identity function h(x) similar to what has been mentioned in the paper?

Second, when looking into the implementation of resnet unit v2, I found this implementation:

	def _building_block_v2(inputs, filters, training, projection_shortcut, strides,
						   data_format):

		shortcut = inputs
		inputs = batch_norm(inputs, training, data_format)
		inputs = tf.nn.relu(inputs)

		# The projection shortcut should come after the first batch norm and ReLU
		# since it performs a 1x1 convolution.
		if projection_shortcut is not None:
			shortcut = projection_shortcut(inputs)

shouldn't we perform projection on the original input before performing BN and RELU? I am asking since this is what I have seen in the second image above. I guess we shouldn't touch the identity/shortcut path...

Please correct me if I'm mistaken.

Any help is much appreciated!!!

Here is the link to the stack overflow question:
https://stackoverflow.com/questions/50766364/residual-unit-implementation-in-tensorflow

The question is just about the implementation details, and why this is the case? I am not trying yet to run the code and do anything yet. Thank!!

Here are some more details:

top-level directory of the model you are using: N/A 
Have I written custom code: N/A
OS Platform and Distribution: Windows 10
TensorFlow installed from: binaries
TensorFlow version: 1.8
Bazel version: N/A
CUDA/cuDNN version: 9.0/7.0
GPU model and memory: GTX 1080 Ti, 11 GB of memory
Exact command to reproduce: N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
What is the top-level directory of the model you are using
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by issa-s-ayoub** [comment]
The question is just about the implementation details, and why this is the case? I am not trying yet to run the code and do anything yet. Thank you for your effort!!

Please note that I have updated my issue above!!

**Comment by robieta** [comment]
The butler is a bot; it probably just fired because your post didn't match some regex.

As to the questions posed.

1) Why is the batch norm placed where it is with respect to projections?
So if I'm reading this correctly you want to know why
```
  if projection_shortcut is not None:
    shortcut = projection_shortcut(inputs)
    shortcut = batch_norm(inputs=shortcut, training=training,
                          data_format=data_format)
  ...
  inputs = batch_norm(inputs, training, data_format)
  inputs += shortcut
```
can't be replaced with:
```
  if projection_shortcut is not None:
    shortcut = projection_shortcut(inputs)
  ...
  inputs = conv2d_fixed_padding(
      inputs=inputs, filters=filters, kernel_size=3, strides=1,
      data_format=data_format)
  inputs += shortcut
  inputs = batch_norm(inputs, training, data_format)
```
The answer gets to the heart of what makes ResNet ResNet. In the case that inputs is zeros (which one somewhat expects for the later layers given the arguments in the ResNet papers) the first implementation is basically a bunch of no-ops, while the second one results in the shortcuts being batch_normed over and over. (Note: batch norm is NOT idempotent) Projections are unusual in that they only occur when the network changes size, and are convolutions in our implementation which is why batch norm is applied there.

2) Why do the V2 shortcuts have batch norm applied to them.
The rationale here is similar to (1). We're applying a learned operation (convolution) during projection, so we should also normalize. And because it's V2, normalization precedes convolution. As an aside, if you don't normalize you can run into numerical issues because the inputs are not necessarily well scaled.

If you don't find those explanations satisfactory feel free to reopen the issue.
P.S. Brownie points for including the diagram images and not making me dig up the papers.
